### PR TITLE
Expose python unsafe buffer pointer

### PR DESCRIPTION
**Pull request by vanbasten23**
Similar to what JAX does https://github.com/openxla/xla/blob/f8cafa2b4ae9d7cf96adca75aaad37952be1a28b/xla/python/py_array.cc#L1598, we can use this helper to make sure dlpack does zero copy, by comparing the buffer point such as https://github.com/google/jax/blob/17d0eb30fa609df2a404543925ca50ae92b5c618/tests/array_interoperability_test.py#L84. This will be used in the followup dlpack PR.

### CODE REVIEW

File: torch_xla/csrc/init_python_bindings.cpp [hunk 1]
```diff
@@ -2465,6 +2465,29 @@ void InitXlaModuleBindings(py::module m) {
...
+            return runtime::GetComputationClient()->UnsafeBufferPointer(data);
+          } else if (xtensor->CurrentIrValue().node != nullptr) {
+            DeviceData* device_data =
+                DeviceData::Cast(xtensor->CurrentIrValue().node.get());
+            if (device_data != nullptr) {
+              torch::lazy::BackendDataPtr data = device_data->data();
+              return runtime::GetComputationClient()->UnsafeBufferPointer(
+                  UnwrapXlaData(data));
+            } else {
+              XLA_ERROR() << "Could not get the buffer pointer.";
```
Comment (JackCaoG) [inline torch_xla/csrc/init_python_bindings.cpp:2485]: it is better to be a bit specified, like 
```
XLA_ERROR() << "Could not get the buffer pointer for XLATensor with IR that's not DeviceData";
```

File: torch_xla/csrc/init_python_bindings.cpp [hunk 2]
```diff
@@ -2465,6 +2465,29 @@ void InitXlaModuleBindings(py::module m) {
...
+            }
+          }
+          XLA_ERROR() << "Could not get the buffer pointer.";
```
Comment (JackCaoG) [inline torch_xla/csrc/init_python_bindings.cpp:2488]: ditto 
```
'Could not get the buffer pointer for XLATensor without XLAData and IR'
```

File: torch_xla/csrc/runtime/pjrt_computation_client.cc [hunk 3]
```diff
@@ -458,6 +458,16 @@ std::vector<ComputationClient::DataPtr> PjRtComputationClient::ReshardData(
   return resharded_results;
 }
 
+std::uintptr_t PjRtComputationClient::UnsafeBufferPointer(
+    const DataPtr handle) {
+  std::shared_ptr<PjRtData> pjrt_data = ReplicateShardedData(handle);
```
Comment (JackCaoG) [inline torch_xla/csrc/runtime/pjrt_computation_client.cc:463]: why do you need to call `ReplicateShardedData`? Is it to handle the SPMD cases?

File: torch_xla/csrc/unwrap_data.h [hunk 4]
```diff
@@ -11,6 +11,9 @@
 
 namespace torch_xla {
 
+runtime::ComputationClient::DataPtr UnwrapXlaData(
```
Comment (JackCaoG) [inline torch_xla/csrc/unwrap_data.h:14]: I didn't see implementation for this function, did I miss anything?

File: torch_xla/csrc/runtime/pjrt_computation_client.cc (hunk 3 above)
Comment (vanbasten23) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:463]: I don't intend to handle the SPMD case atm. Let me change it.

File: torch_xla/csrc/unwrap_data.h (hunk 4 above)
Comment (vanbasten23) [reply torch_xla/csrc/unwrap_data.h:14]: The implementation is at https://github.com/pytorch/xla/blob/2907ab3093753d8a3516c8f67337a9464bf4ed8a/torch_xla/csrc/unwrap_data.cpp#L14. Somehow it's missing in the header..

**Review by JackCaoG — APPROVED** [review]

**Comment by vanbasten23** [comment]
Thanks Jack for the review!

**Pull request merged**